File: day12/day12.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <string>

using namespace std;

struct vec
{
	int x;
	int y;
};

static vec xmul(const vec v1, const vec v2)
{
	return {
		v1.x * v2.x - v1.y * v2.y,
		v1.x * v2.y + v1.y * v2.x
	};
}

static int manhattan(const vec v)
{
	return (v.x > 0 ? v.x : -v.x) +	(v.y > 0 ? v.y : -v.y);
}

static vec operator+(const vec v1, const vec v2)
{
	return { v1.x + v2.x, v1.y + v2.y};
}

static vec operator*(const vec v1, int n)
{
	return { v1.x * n, v1.y * n };
}

static const vec north = {0, -1 };
static const vec south = {0, 1 };
static const vec east = {1, 0};
static const vec west = {-1, 0};
static const vec rotleft = {0, -1 };
static const vec rotright = {0, 1};

static int navigate_part1(const vector<string>& instructions)
{
	vec pos = {0, 0};
	vec dir = east;
	for (auto& ins: instructions)
	{
		int val = stoi(ins.substr(1));
		switch (ins[0])
		{
		case 'N': pos = pos + north * val; break;
		case 'S': pos = pos + south * val; break;
		case 'E': pos = pos + east * val; break;
		case 'W': pos = pos + west * val; break;
		case 'F': pos = pos + dir * val; break;
		case 'L':
			val /= 90;
			while (val-- > 0)
			{
				dir = xmul(dir, rotleft);
			}
			break;
		case 'R':
			val /= 90;
			while (val-- > 0)
			{
				dir = xmul(dir, rotright);
			}
			break;
		}
	}

	return manhattan(pos);
}

static int navigate_part2(const vector<string>& instructions)
{
	vec pos = {0, 0};
	vec wp = east * 10 + north * 1;
	for (auto& ins: instructions)
	{
		int val = stoi(ins.substr(1));
		switch (ins[0])
		{
		case 'N': wp = wp + north * val; break;
		case 'S': wp = wp + south * val; break;
		case 'E': wp = wp + east * val; break;
		case 'W': wp = wp + west * val; break;
		case 'F': pos = pos + wp * val; break;
		case 'L':
			val /= 90;
			while (val-- > 0)
			{
				wp = xmul(wp, rotleft);
			}
			break;
		case 'R':
			val /= 90;
			while (val-- > 0)
			{
				wp = xmul(wp, rotright);
			}
			break;
		}
	}

	return manhattan(pos);
}
int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << "Usage: " << argv[0] << " <filename>" << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << "Cannot open " << argv[1] << endl;
		return 1;
	}

	vector<string> instructions;
	string line;
	while (getline(input, line))
	{
		instructions.emplace_back(move(line));
	}
	input.close();

	cout << "Part1: " << navigate_part1(instructions) << endl
	     << "Part2: " << navigate_part2(instructions) << endl;
	return 0;
}

File: day12/day12.py
#!/usr/bin/env python3

def xmul(a, b):
    return (a[0]*b[0]-a[1]*b[1], a[0]*b[1]+a[1]*b[0])

def add(a, b):
    return (a[0]+b[0], a[1]+b[1])

def nmul(a, b):
    return (a[0] * b, a[1] * b)

def manhattan(a):
    return abs(a[0]) + abs(a[1])

NORTH = (0,  -1)
SOUTH = (0,  1)
EAST  = (1,  0)
WEST  = (-1, 0)
LEFT  = (0,  -1)
RIGHT = (0,  1)

def navigate_part1(txt):
    curdir = EAST
    curpos = (0, 0)
    for line in txt.strip().split("\n"):
        ins = line[0]
        n = int(line[1:])
        if ins == "N":
            curpos = add(curpos, nmul(NORTH, n))
        elif ins == "S":
            curpos = add(curpos, nmul(SOUTH, n))
        elif ins == "E":
            curpos = add(curpos, nmul(EAST, n))
        elif ins == "W":
            curpos = add(curpos, nmul(WEST, n))
        elif ins == "F":
            curpos = add(curpos, nmul(curdir, n))
        elif ins == "L":
            for j in range(n//90):
                curdir = xmul(curdir, LEFT)
        elif ins == "R":
            for j in range(n//90):
                curdir = xmul(curdir, RIGHT)
        else:
            pass

    return manhattan(curpos)

def navigate_part2(txt):
    waypoint = add(nmul(EAST, 10), nmul(NORTH, 1))
    curpos = (0, 0)
    for line in txt.strip().split("\n"):
        ins = line[0]
        n = int(line[1:])
        if ins == "N":
            waypoint = add(waypoint, nmul(NORTH, n))
        elif ins == "S":
            waypoint = add(waypoint, nmul(SOUTH, n))
        elif ins == "E":
            waypoint = add(waypoint, nmul(EAST, n))
        elif ins == "W":
            waypoint = add(waypoint, nmul(WEST, n))
        elif ins == "F":
            curpos = add(curpos, nmul(waypoint, n))
        elif ins == "L":
            for j in range(n//90):
                waypoint = xmul(waypoint, LEFT)
        elif ins == "R":
            for j in range(n//90):
                waypoint = xmul(waypoint, RIGHT)

    return manhattan(curpos)

txt = """F10
N3
F7
R90
F11
"""
assert navigate_part1(txt) == 25
assert navigate_part2(txt) == 286

with open("input", "rt") as f:
    txt = f.read()

print("Part1:", navigate_part1(txt))
print("Part2:", navigate_part2(txt))
